### PR TITLE
Join stake - fee boundary values tests

### PR DESCRIPTION
**Pull request by piotr-iohk**
# Issue Number

#896

# Overview

- [x] I have added some integration tests for fee boundary values
- [ ] adjusted linear_fee in genesis.yaml to use also `certificate`

# Comments
**In comments:** questions/suggestions about adjusting error messaging

### CODE REVIEW

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs [hunk 1]
```diff
@@ -207,6 +209,64 @@ spec = do
...
+            w <- fixtureWalletWith ctx [fee]
+            r <- joinStakePool ctx (p ^. #id) (w, "Secure Passphrase")
+            print r
+            expectResponseCode HTTP.status403 r
+            expectErrorMessage errMsg403Fee r
+            -- the error msg here is a bit misleading:
+            -- I'm unable to adjust the given transaction to cover the associated fee!
+            -- In order to do so, I'd have to select one or more additional inputs,
+            -- but I can't do that without increasing the size of the transaction
+            -- beyond the acceptable limit. Note that I am only missing 42 Lovelace.
```
Comment (KtorZ) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: Hmmm. That's an interesting case which could deserve some more investigation.

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs [hunk 2]
```diff
@@ -207,6 +209,64 @@ spec = do
...
+            -- IMO should be changed to smth like:
+            -- "Cannot join stake-pool!
+            -- I have enough ada to cover fee in order to delegate
+            -- but 0 ada to delegate after fee is deducted"
+
+        it "STAKE_POOLS_JOIN_01x - \
+            \I cannot join if I have not enough fee to cover" $ \ctx -> do
+            (_, p:_) <- eventually $
+                unsafeRequest @[ApiStakePool] ctx listStakePoolsEp Empty
+            let fee = 42
```
Comment (KtorZ) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: **[suggestion]**

Move this constant to be a named test data.

**Why**

Avoid magic numbers: https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#prefer-named-constants-over-magic-numbers

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs (hunk 1 above)
Comment (piotr-iohk) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: My understanding was that the behavior is correct, cause it doesn't make sense to delegate 0 funds. In that case adjusting error message should be enough... (?)

Comment (KtorZ) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: It's not really about delegating 0 funds. When you create the delegation certificates, it's not like taking a snapshot of your current funds and delegating them. It's simply creates a note that tells the node that any address containing your reward account (grouped address) should be delegated to a particular pool. So, all addresses, before and after the certificate submission carrying a ref to the reward account will be delegated. 

There's no problem to create a certificate and have no funds to delegate _yet_.

**Comment by paweljakubas** [comment]
@piotr-iohk I second all remarks raised by @KtorZ I would add more : 
1. Please rebase when quitStakePool PR is merged
2. Please consider checking also these cases in the context of quitStakePool. Here, we also form tx and money might be lacking to cover fee. So in these cases, you need to have money to cover joining, and then quiting (keep this in mind)
3. If those tests are passing and you are satisfied please think about setting join/quit endpoint in swagger as *Stable*

**Comment by paweljakubas** [comment]
@piotr-iohk #1080 merged into master (see point 1 above)

File: lib/jormungandr/test/data/jormungandr/genesis.yaml [hunk 3]
```diff
@@ -49,7 +49,7 @@ blockchain_configuration:
   linear_fees:
     constant: 42
     coefficient: 0
-    certificate: 0
+    certificate: 1
```
Comment (KtorZ) [inline lib/jormungandr/test/data/jormungandr/genesis.yaml:52]: This is actually interesting. We should probably use some non-null fee for the coefficient part as well.

**Comment by KtorZ** [comment]
Okay I've moved the business logic about joining/quitting stake pools inside the wallet engine and actually defined dedicated error types for these inside of overpopulating the `ErrSelectForDelegation`!

I've also revised a bit the error message, in particular when there's not enough funds to pay for the delegation certificate (either joining or quitting). 

Also caught a bug and had to adjust the return type for API transaction outputs: they can be empty! In the context of delegation this makes perfect sense (in particular, in the testing scenario, this can happen if you select just the right amount, not change and no output).

**Review by KtorZ — APPROVED** [review]
